apps/email/js/html_cache_restore.js

```diff
@@ -21,23 +66,143 @@ var startupCacheEventsSent = false;
+      // since the worker will have started up and registered to get those
+      // messages. It is OK though since not showing any UI. If this case is
+      // triggered, it will result in a UI display in that case, but that is OK,
+      // it only happens on an app update situation, from cookies to
+      // localStorage or some kind of localStorage reset. So a rare event, and
+      // does not cause harm, just a bit of extra work in those cases once.
+      console.log('localOnModelLoaded called, hasAccount: ' +
+                  localStorage.getItem('data_has_account'));
+
+      setDefaultView();
+      hydrateHtml(startupData.view);
+      startupOnModelLoaded = null;
+      callback();
+    });
+  }
+
+  if (!localStorage.getItem('data_has_account')) {
+    console.log('data_has_account unknown, asking for model load first');
+    startupOnModelLoaded = localOnModelLoaded;
+  }
+
+  function hasAccount() {
+    // var _1 = performance.now();
+    var has = localStorage.getItem('data_has_account') === 'yes';
+    // console.log('@@@ LOCALSTORAGE GET data_has_account: ' +
+    //             (performance.now() - _1));
+    return has;
+  }
+
+  function setDefaultView() {
+    if (hasAccount()) {
+      if (!startupData.view) {
+        startupData.view = 'message_list';
+      }
+    } else {
+      startupData.view = 'setup_account_info';
+    }
+  }
+
+  // Set up the default view, but if that is not possible to know yet, since
+  // the status of hasAccount is unknown, wait for the callback to set it up.
+  if (!startupOnModelLoaded) {
+    setDefaultView();
+  }
+
+  startupData.entry = 'default';
+
+  /*
+  function parseFragmentId() {
+    var data = {},
+        frag = location.href.split('#')[1];
+
+    if (frag) {
+      frag.split('&').forEach(function(nameValue) {
+        var nvParts = nameValue.split('=');
+        data[decodeURIComponent(nvParts[0])] = decodeURIComponent(nvParts[1]);
+      });
+    }
+
+    return data;
+  }
+  */
+
+  /**
+   * Makes sure messages are not coming in at a fast rate due to things like
+   * double clicks. Only necessary to use if the message is something that would
+   * insert cards, and fast entries could mess up card state.
+   */
+  function messageTooFast(type) {
+    // First check for fast double taps
+    var entryTime = Date.now();
+    // Only one entry per second, but request-sync messages are OK, as they
+    // are background tasks. The entryTime is about avoiding messing up the UI
+    // startup from double-click taps of activities/notifications that would
+    // insert new cards.
+    if (entryTime < lastEntryTime + 1000) {
+      console.log('email entry gate blocked fast repeated action: ' + type);
+      return true;
+    }
+
+    lastEntryTime = entryTime;
+    return false;
+  }
+
+  /**
+   * If waiting for the first startup message, only accept the message if its
+   * type matches the pending expected type.
+   */
+  function ignoreMessage(type) {
```

Comment:
So, it's not the right name, but I think something like "acquireSystemMessageSemaphore" is more appropriate.  Maybe something involving a "gate"?  Basically, I understand this function for being responsible for:

- Telling the other less important message types to get lost because they're being ignored by returning false.
- Letting the message type we decided we do want to process through, and taking down the "get lost" sign on our way in.

It might make sense to introduce a function shouldProcessMessage() that internally uses messageTooFast and ignoreMessage.  As noted elsewhere, it's currently possible to cause problems by mis-ordering ignoreMessage and messageTooFast, so this makes an arguably more clear contract.